### PR TITLE
Kubectl: Plumb openapi validation (disabled by default)

### PR DESCRIPTION
**Pull request by apelisse**
**What this PR does / why we need it**: Creates a new flag '--openapi' and plumb in the validation code so that it can be used by default to validate objects against the openapi schema.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: partially https://github.com/kubernetes/kubectl/issues/49

**Special notes for your reviewer**:

This is not complete, the name of the variable must change for example.

**Release note**:
```release-note
Kubectl uses openapi for validation. If OpenAPI is not available on the server, it defaults back to the old Swagger.
```

### CODE REVIEW

**Comment by apelisse** [comment]
@mengqiy have you seen the error? Do you know what's going on?

**Comment by apelisse** [comment]
OK, We can't see the actual errors from the given link. I've reproduced locally, fixing ...

**Comment by mengqiy** [comment]
>OK, We can't see the actual errors from the given link. I've reproduced locally, fixing ...

Not sure what link are you referring to.
Looks the tests are all passing. I assume you have fixed the issue.

**Comment by apelisse** [comment]
Yeah, I have, feel free to take a look!

File: pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 1]
```diff
@@ -402,8 +404,20 @@ func (f *ring1Factory) AttachablePodForObject(object runtime.Object, timeout tim
 	return pod, err
 }
 
-func (f *ring1Factory) Validator(validate bool, cacheDir string) (validation.Schema, error) {
+func (f *ring1Factory) Validator(validate bool, openapi bool, cacheDir string) (validation.Schema, error) {
```
Comment (mengqiy) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: nit: `validate, openapi bool`

File: pkg/kubectl/cmd/util/openapi/validation/validation_test.go [hunk 2]
```diff
@@ -41,8 +43,30 @@ var _ = Describe("resource validation using OpenAPI Schema", func() {
 		Expect(validator).ToNot(BeNil())
 	})
 
+	It("finds and validates a deployment", func() {
```
Comment (mengqiy) [inline pkg/kubectl/cmd/util/openapi/validation/validation_test.go]: > finds and validates a deployment

I don't understand what does "find" mean here.

**Review by mengqiy — COMMENTED** [review]
2 nits

**Comment by mengqiy** [comment]
/lgtm
Feel free to re-apply the label after addressing the nits :)

**Comment by apelisse** [comment]
@mengqiy I've addressed your feedback (I think). I've also changed the name of the command-line argument (from `openapi` to `openapi-validation` to be more specific) and it's default from `true` to `false`.

I don't think it makes sense to have it enabled by default until #50404 gets merged. I'll make another PR to revert that value once the cache is "enabled".

**Comment by mengqiy** [comment]
> I've also changed the name of the command-line argument (from openapi to openapi-validation to be more specific) and it's default from true to false.

I'm using `--openapi` for  kubectl apply as well.
But if you change the name, I can't reuse it in apply. I will need to add a separate flag.

WDYT if we have one flag gating all openapi related behavior?

**Comment by apelisse** [comment]
> WDYT if we have one flag gating all openapi related behavior?

@pwittrock Do you have an opinion on that?

**Comment by apelisse** [comment]
@pwittrock can you please approve this?

**Pull request merged**